cmd/jaeger/internal/integration/e2e/integration.go

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	jaeger2otlp "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/jaeger"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/testbed/testbed"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+
+	"github.com/jaegertracing/jaeger/cmd/jaeger/internal"
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/plugin/storage/integration"
+	"github.com/jaegertracing/jaeger/ports"
+	"github.com/jaegertracing/jaeger/proto-gen/api_v2"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+var (
+	_ spanstore.Writer = (*spanWriter)(nil)
+	_ spanstore.Reader = (*spanReader)(nil)
+)
+
+// E2EStorageIntegration holds components for e2e mode of Jaeger-v2
+// storage integration test. The intended usage is as follows:
+//   - A specific storage implementation declares its own test functions
+//     (e.g. starts remote-storage).
+//   - In those functions, instantiates with e2eInitialize()
+//     and clean up with e2eCleanUp().
+//   - Then calls RunTestSpanstore.
+type E2EStorageIntegration struct {
+	integration.StorageIntegration
+	ConfigFile string
+
+	runner        testbed.OtelcolRunner
+	configCleanup func()
+}
+
+// e2eInitialize starts the Jaeger-v2 collector with the provided config file,
+// it also initialize the SpanWriter and SpanReader below.
+func (s *E2EStorageIntegration) e2eInitialize() error {
```

Comment:
please use the pattern of passing `*testing.T` to all functions and doing `require.NoError(t, err)` instead of returning errors. This makes the code cleaner and easier to debug.

Reply:
except for internals of span reader / writer, those are better as proper (non-test) components.